### ISSUE TITLE
Unnecessary strictNullChecks errors for nested property access protected by type predicate

### ISSUE THREAD

**Issue by ashsearle** [microsoft/TypeScript#50566]
# Bug Report

TypeScript shows strictNullChecks errors for "possible 'null'" and "possible 'undefined'" values along a property access chain, even though the code is protected by a type predicate using optional chaining that guarantees the property access chain is valid.

### 🔎 Search Terms

NonNullable type predicate optional chaining

### 🕗 Version & Regression Information

- This is the behavior in every version I tried (from 3.7.5. to 4.9.0 nightly), and I reviewed the FAQ for entries about type system behaviour, type guards and save navigation operator.

### ⏯ Playground Link

**Playground Link:** [Provided](https://www.typescriptlang.org/play?ts=4.9.0-dev.20220831#code/C4TwDgpgBAyg9gWwgFXNAvFA3gKClMAJzjAH4AubPfKAOwgGdgIATABWLMqcIEtaA5tQC+OUSwgBjADYBDQtABmAV1qTgvOLSgNEEAEqzaLRADFVkgBQA6W-IENKRkAG0AugEpK8JKkhQAHzplaWkAbhxJLSYoXgYAEQhFflYAQWMAOThgDJDpKEwAHmQAGgA+SwA3WWlKZEDg0IbVCWT6Fi8oavy4qCzaXNDZACNpCGKygsnuqABCTFo8iMjo4FjaJiNJDB09Q2MzC0sPZd5FKEs4xLa0zOzB6UuN4C2IUmsiEnf6JlYOEg8Hio+AA9CCoABVegAD0g6lYOmAfHUDwAwgALKQAawYUAghGIhFxWigwExUGkKSgwwg0jgAHdyNQwVAALTrTZqCCUADywwAVlI1r0wHAGAxeKMQFAAOSLUIy5ng9n8TnbXkCoWxXGi8WS6TSmUtJIpFiK0HKgicDWC9TaghiiVS2XGm5m6i6JD7EwIcxqJ5qiAfTjWH7MdicE5iHA4M4XahXE3tdIsLI5PIBl5coEAMhzCYSSduqfuGdVWe270+YFz+fwibdKbTD0zryrnG+jHD-xr1CBuAtfTgiOR6dCGOxuPxhNxmIUJWpyjWwDgw4Qykk6Kgqno23F8mlUQkUBGcEqEA9eyMPr9VnLr2DJFDXb+kYioiAA)

### 💻 Code

```ts
type SomeType = {
  prop?: {
    nestedProp?: string
  }
}
declare function someRandomFunc(...args: any[]): SomeType | null;
const isDefinedAndNotNull = <T,>(val: T | null | undefined): val is NonNullable<T> => val != null;

const instance = someRandomFunc();

if (isDefinedAndNotNull(instance?.prop?.nestedProp)) {
  // Unexpected strictNullChecks errors on the line below:
  // - instance: Object is possibly 'null'
  // - instance: Object is possibly 'undefined'
  // - prop: Object is possibly 'undefined'
  someRandomFunc(instance.prop.nestedProp);
}

if (
  isDefinedAndNotNull(instance) &&
  isDefinedAndNotNull(instance?.prop) &&
  isDefinedAndNotNull(instance?.prop?.nestedProp)
  ) {
  // No strictNullChecks errors here, but too much unnecessary code above
  someRandomFunc(instance.prop.nestedProp);
}

```

<details><summary><b>Compiler Options</b></summary>

```json
{
  "compilerOptions": {
    "strict": true,
    "noImplicitAny": true,
    "strictNullChecks": true,
    "strictFunctionTypes": true,
    "strictPropertyInitialization": true,
    "strictBindCallApply": true,
    "noImplicitThis": true,
    "noImplicitReturns": true,
    "alwaysStrict": true,
    "esModuleInterop": true,
    "declaration": true,
    "experimentalDecorators": true,
    "emitDecoratorMetadata": true,
    "target": "ES2017",
    "jsx": "react",
    "module": "ESNext",
    "moduleResolution": "node"
  }
}
```

</details>

### 🙁 Actual behavior

TypeScript emits errors about possible null / undefined values along a property access chain.

This is unexpected as the code is guarded by a type predicate. 

### 🙂 Expected behavior

No strictNullChecks errors.

**Comment by fatcerberus** [comment]
This seems like an expected limitation, given that checking the value of a property doesn't narrow the containing object in general (except for discriminated unions).  I note that there are potentially multiple steps of inference required:

* `instance?.prop?.nestedProp` is known to **not** be `undefined`, therefore
* `instance?.prop` must also be non-nullish, therefore
* `instance` must also be non-nullish

It would indeed be nice if this pattern worked, but I don't know if the control-flow analysis machinery as it exists today is capable of making these inferences.

**Comment by MartinJohns** [comment]
Duplicate of #34974.

**Comment by fatcerberus** [comment]
As pointed out in the dupe'd issue, this works if you write the type check inline:
```ts
if (instance?.prop?.nestedProp != null) {
  someRandomFunc(instance.prop.nestedProp);  // saul goodman
}
```

So ignore my comment above - TS *can* do the necessary inferences, it's just that `NonNullable<T>` is opaque to it.

**Comment by ashsearle** [comment]
Thanks for the comments. @RyanCavanaugh's comments on the original issue gives me hope this might be fixed in the future, and there's no need for me to leave this duplicate issue open.
